### PR TITLE
Create workflow to publish to npm

### LINKED ISSUE #4: Set up CI release publishing workflow
We should create a GitHub Actions workflow to publish releases to npm under the Azavea account. Releases will be semantically versioned and tagged on `main`.

We should also create a release template that details how to version and tag a release so that it will correctly trigger this workflow.

### PR DESCRIPTION
**Pull request by ColeKettler**
## Overview

Establishes a GitHub Actions workflow to build and publish `react-showtime` to npm.

Resolves #4 

## Testing Instructions

- Verify that the [Netlify build](https://app.netlify.com/sites/react-showtime/deploys/5fb7eed9c976910007ab32cf) passes.
- Verify that the [deploy preview](https://deploy-preview-7--react-showtime.netlify.app/) is available with no content security policy errors in the console.
- Verify that the [CI workflow passes](https://github.com/azavea/react-showtime/pull/7/checks?check_run_id=1432080802), including all linting and tests.
- Confirm that all four files in `dist` will be included in the final released package as part of the ["Build and publish package" step in the Release workflow](https://github.com/azavea/react-showtime/pull/7/checks?check_run_id=1432080568).

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/RELEASE.md
Comment (ColeKettler) [inline .github/ISSUE_TEMPLATE/RELEASE.md:12]: After discussing with Scott, we determined that our typical Git Flow workflow would be most comfortable here.

File: .github/ISSUE_TEMPLATE/RELEASE.md
Comment (ColeKettler) [inline .github/ISSUE_TEMPLATE/RELEASE.md:16]: Yarn will create an annotated tag with the `v` prefix by default. We suppress this by using `--no-git-tag-version`, allowing Git Flow tooling to take over the tagging process.

File: .github/ISSUE_TEMPLATE/RELEASE.md
Comment (ColeKettler) [inline .github/ISSUE_TEMPLATE/RELEASE.md:34]: This will kick off the `release.yml` workflow.

File: .github/workflows/continuous_integration.yml
Comment (ColeKettler) [inline .github/workflows/continuous_integration.yml:1]: I decided to go with a separate CI workflow rather than overloading Netlify. Currently our Netlify build doesn't lint or run tests on the `react-showtime` module. We could do that by expanding on our build script and / or `package.json` hooks, but I don't think it provides feedback for a PR as clearly or offers as much configuration flexibility as GitHub Actions does. I would prefer keeping Netlify focused on deploying the static site, and Actions focused on testing and publishing the library.

File: .github/workflows/continuous_integration.yml
Comment (ColeKettler) [inline .github/workflows/continuous_integration.yml:9]: Neat thing that Rocky introduced to me here. This creates an option to [trigger the workflow manually](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/) if needed. Could come in handy.

File: .github/workflows/continuous_integration.yml
Comment (ColeKettler) [inline .github/workflows/continuous_integration.yml:14]: Actions is throwing some warnings that this will be updated from `18.04` to `20.04` soon. Good to keep in mind, but I don't think it's a huge risk here.

File: .github/workflows/continuous_integration.yml
Comment (ColeKettler) [inline .github/workflows/continuous_integration.yml:21]: Turns out the default Ubuntu runner includes `nvm`! You need to set up a login shell for it to work, but it takes for the rest of the job. Takes 2 - 3 seconds to install the correct version.

File: .github/workflows/continuous_integration.yml
Comment (ColeKettler) [inline .github/workflows/continuous_integration.yml:24]: Yarn has utilities included to [install and enforce which version of Yarn you use](https://classic.yarnpkg.com/en/docs/cli/policies/). I think `yvm` is more convenient for local dev, but `policies` works well here. Under the hood, it's installing Yarn and saving a bundled copy of it in a local directory. If we based our workflow around this we'd commit that copy and Yarn would know to use it automatically. In this context, it just makes it easy to quickly install the correct version of Yarn.

File: .github/workflows/continuous_integration.yml
Comment (ColeKettler) [inline .github/workflows/continuous_integration.yml:37]: This is a little funky because the location of the Yarn cache varies between OS and which version of Yarn you're using. This is [the approach recommended](https://github.com/actions/cache/blob/master/examples.md#node---yarn) by the `cache` Action.

File: .github/workflows/continuous_integration.yml
Comment (ColeKettler) [inline .github/workflows/continuous_integration.yml:40]: Rough equivalent of `npm ci`.

File: .github/workflows/release.yml
Comment (ColeKettler) [inline .github/workflows/release.yml:6]: In place for testing this PR, will be removed prior to merge.

File: .github/workflows/release.yml
Comment (ColeKettler) [inline .github/workflows/release.yml:25]: This sets up an `.npmrc` file to assist in authenticating to npm. By excluding `node_version`, we defer to whichever version of Node is in `PATH`, which will be managed by `nvm` here.

File: .github/workflows/release.yml
Comment (ColeKettler) [inline .github/workflows/release.yml:36]: I regenerated a publish token in our npm account. It has been saved as a secret and made available to this repo.

File: .github/workflows/release.yml
Comment (ColeKettler) [inline .github/workflows/release.yml:34]: This is here for testing and will output the files that will be included in the final package. Yarn unfortunately lacks a `--dry-run` option for its publish command. I'd like to perform the actual release post-merge to exercise the release instructions included in an issue template. I have a high degree of confidence that `yarn publish` will just work, but I can try to create a test case to increase our confidence if necessary.

File: CHANGELOG.md
Comment (ColeKettler) [inline CHANGELOG.md:16]: This tag doesn't exist yet but it will be created right after this PR is merged.

File: example/package.json
Comment (ColeKettler) [inline example/package.json:13]: This will be changed back to `develop` before merging, and changed to `0.1.0` as part of the subsequent release.

File: package.json
Comment (ColeKettler) [inline package.json:14]: This establishes [an allow list](https://docs.npmjs.com/cli/v6/configuring-npm/package-json#files) for files to be included in the final package. I've verified that these files are all that are necessary to run `react-showtime`.

File: CHANGELOG.md
Comment (ColeKettler) [inline CHANGELOG.md:7]: This will be rotated to `0.1.0` as part of the release that will happen after this PR is merged.

File: example/package.json
Comment (ColeKettler) [inline example/package.json:13]: It took a while to arrive back where we started here, but I think linking to the local version of `react-showtime` is the best overall approach. Since we've decided to use Git Flow, all merges to `main` will happen as part of a release process, and as such all commits to `main` will be tagged. The production example site will effectively always use the most recent stable release of `react-showtime`.

Going with this approach means we don't have to worry about starting a hotfix branch (or another release) to apply updates to the example site post-release and we don't need to worry about continually bumping `example/yarn.lock`. It also cuts the size of the release checklist in half and takes the manual `link` steps out of local development. This satisfies all of our requirements given our workflow.

File: .github/workflows/continuous_integration.yml
Comment (hectcastro) [reply .github/workflows/continuous_integration.yml:1]: Sounds good.

File: .github/workflows/continuous_integration.yml
Comment (hectcastro) [reply .github/workflows/continuous_integration.yml:9]: Yep, I also got exposure to this via: https://github.com/azavea/scala-steward/blob/e75d8ffdb3556fed2834af413f6f1162efb88bca/.github/workflows/scala-steward.yml#L3-L6

File: .github/workflows/continuous_integration.yml
Comment (hectcastro) [reply .github/workflows/continuous_integration.yml:21]: For the context of CI, it seems more important to be able to support execution against a range of Node versions than the one the project is specifically pinned to. That makes me more inclined to pursue a `setup-node` Action approach with a pin on version 12.x to start.

What do you think?

File: .github/workflows/continuous_integration.yml
Comment (hectcastro) [reply .github/workflows/continuous_integration.yml:24]: Similar comment here regarding Yarn -- although I'm not sure what version gets packaged through `setup-node` or the base image.

File: .github/workflows/release.yml
Comment (hectcastro) [reply .github/workflows/release.yml:25]: Similar comment here regarding use of `nvm` vs. allowing `setup-node` to do its thing with a lose pin on the LTS (e.g., 12.x).

File: .github/workflows/release.yml
Comment (hectcastro) [reply .github/workflows/release.yml:36]: Minor nitpick here, but could we rename the variable to include a reference to the NPM user the package will get published under? This is the convention we're currently using with PyPi: `PYPI_AZAVEA_USERNAME`

**Review by hectcastro — COMMENTED** [review]
Looks good. Thanks for all of the in-context comments.

I left a few comments myself, but most hinge on the use of `nvm` vs. depending on `setup-node` within the context of CI.

File: .github/workflows/release.yml
Comment (ColeKettler) [reply .github/workflows/release.yml:36]: For sure, I'll do that.

File: .github/workflows/continuous_integration.yml
Comment (ColeKettler) [reply .github/workflows/continuous_integration.yml:21]: Good point, this is a library and not just an application. I think pinning the publish to a static version makes sense from the perspective of reproducible builds, but tests should be run against multiple supported versions. I think active LTS versions are a good starting point.

File: .github/workflows/continuous_integration.yml
Comment (ColeKettler) [reply .github/workflows/continuous_integration.yml:24]: I'd be comfortable relying on whatever crops up by default for CI builds. It's been fairly stable with respect to installing packages from npm, which is how users will typically consume this library. Pinning the version used to build the published artifact is more valuable than the version used to install it.

File: .github/workflows/continuous_integration.yml
Comment (ColeKettler) [reply .github/workflows/continuous_integration.yml]: After thinking about this, I feel less strongly about pinning the publish process than I originally thought. We can easily see which versions were used in the build in case we need to reproduce failure, and we're exercising the version we will use to publish in the CI workflow. I'm going to change the release workflow to use `setup_node` and whichever version of Yarn comes with it as well. `.nvmrc` and `.yvmrc` can just be for development convenience and consistency.

File: .github/workflows/continuous_integration.yml
Comment (ColeKettler) [inline .github/workflows/continuous_integration.yml:17]: These are all of the currently active LTS versions, and the most recent stable version.

File: .nvmrc
Comment (ColeKettler) [inline .nvmrc:1]: This is the most recent LTS version as of October 27.

**Comment by ColeKettler** [comment]
@hectcastro Ready for the next round of review!

**Review by hectcastro — APPROVED** [review]
Please drop a message in `#operations` with how the publishing goes after `--dry-run` gets removed. 🚀

**Pull request merged**
